internal/typeinfo/scan.go

```diff
@@ -0,0 +1,25 @@
+package typeinfo
+
+import "reflect"
+
+// ScanProxy is a shim for scanning query results
+// into types for which we have information.
+type ScanProxy struct {
```

Comment:
This feels like it's doing too much. I'm not suggesting fixing it now, as you're doing a lot of mechanical changes, but we should have a `MapSetterProxy` and a `ValueSetterProxy`. The `AddScanTarget` should take and return a `Proxy` interface.

```go
type Proxy interface {
    SetValue()
}
```

That way we remove the need for the if statement in `OnSuccess`, in addition to changing the name which doesn't tell you the intent of what it's actually doing.

Reply:
I'm trying to limit this to a lift and shift. 

I actually started to do some other rework, and the change-set got unwieldy very fast. Once we have things a little better organised, I imagine we'll reconsider these implementations wholesale in the context of the features we need to add.

Reply:
I think whether to dispatch or pattern match (see my general comment) comes down to personal opinion in the end, but I also think no paradigm should be absolute. As I said, my fear is that we do not take into account the original design and its quest for "simplicity".

In this concrete case we would be creating one file, one interface, three struct types and three methods to refactor a <20 line function. Is there a reason for that, are we envisioning that we will support more data types or that we need this added complexity?

Reply:
> That way we remove the need for the if statement in OnSuccess, in addition to changing the name which doesn't tell you the intent of what it's actually doing.

We would not be removing the if though, we would be using method dispatch do it instead, effectively we would only be moving the conditional elsewhere.